### ISSUE TITLE
Event subprocess instance gets passed wrong variable values (possible race condition?)

### ISSUE THREAD

**Issue by allanbarklie** [camunda-cloud/zeebe#7064]
**Describe the bug**

If multiple instances of a non interrupting event subprocess are created in quick succession input variables are not passed correctly.

If each start message is given different message variables these are not reliably passed to the separate sub process instances, instead some will receive values intended for a different instance.

So if I try to create 2 event subprocesses one with message="A" and one with message="B" the broker may create two sub processes with message="A"  or 2 with message="B" .

**To Reproduce**

Please see simple attached workflow.
[sub-process-test.bpmn.txt](https://github.com/camunda-cloud/zeebe/files/6502080/sub-process-test.bpmn.txt)

Steps to reproduce the behavior

Rename the attached file to sub-process-test.bpmn

```
./zbctl --insecure deploy sub-process-test.bpmn
./zbctl --insecure create instance sub-process-test --variables '{"correlationValue":1}'
```

In a separate shell window
`./zbctl --insecure create worker print-message --handler cat`

In the original shell
```
./zbctl --insecure publish message "sub-process-started" --correlationKey='1' --variables '{"mes
sage": "A"}' && ./zbctl --insecure publish message "sub-process-started" --correlationKey='1' --variables '{"message": "
B"}'
```

also try with a sleep added between the publish steps

```
./zbctl --insecure publish message "sub-process-started" --correlationKey='1' --variables '{"message": "A"}' && sleep 1 && ./zbctl --insecure publish message "sub-process-started" --correlationKey='1' --variables '{"
message": "B"}'
```

**Expected behavior**

In both cases each handler instance should  log the correct message - something like this:
```
2021/05/18 15:38:59 Activated job 2251799813693657 with variables {"message":"A","correlationValue":1}
2021/05/18 15:38:59 Handler completed job 2251799813693657 with variables {"message":"A","correlationValue":1}
2021/05/18 15:39:00 Activated job 2251799813693670 with variables {"message":"B","correlationValue":1}
2021/05/18 15:39:00 Handler completed job 2251799813693670 with variables {"message":"B","correlationValue":1}
```

However instead in the first case (with no sleep command) we see something like this:

```
2021/05/18 15:37:47 Activated job 2251799813693364 with variables {"message":"B","correlationValue":1}
2021/05/18 15:37:47 Handler completed job 2251799813693364 with variables {"message":"B","correlationValue":1}
2021/05/18 15:37:47 Activated job 2251799813693368 with variables {"message":"B","correlationValue":1}
2021/05/18 15:37:47 Handler completed job 2251799813693368 with variables {"message":"B","correlationValue":1}
```

(Or it might print A twice)

</details>

**Environment:**
 - OS: WSL2  (Centos8)
 - Zeebe Version: 0.26.3
 - Configuration: hazelcast/simple monitor/zeeqs, single broker with no separate gateway

**Comment by saig0** [comment]
@allanbarklie thanks for raising this up.

In the example, the variable `message` is overridden by the second message before the logging task is executed. This happens because the variable is defined in the (root) scope of the process instance. 

It can be solved by defining an input mapping on the **event subprocess**: 

```
message

= null
```

Does this help you?

Unfurtunatelly, this is not mentioned in the [docs](https://docs.camunda.io/docs/reference/bpmn-processes/event-subprocesses/event-subprocesses#variables) explicitly. Feel free to contribute and improve the docs :rocket:

**Comment by allanbarklie** [comment]
@saig0 
Thanks for your patience and explanation.
I think I now understand the current intended behaviour, as you say there doesn't seem to be anything in the docs that describes this or makes this obvious - hopefully we can make that better.

I'd like to check my understanding, are the following descriptions correct?

Generally when creating elements or activities that introduce parallelism care needs to taken to understand the scope and sharing of variables.

Specifically by default all elements and activities within a parallel execution scope continue to access variables that are shared with all other parallel execution scopes.  This explicitly includes:
Any i/o mapping performed in start events such as those found in non interrupting event sub processes and multi-embedded subprocesses.
Any variables sent in messages to events within such parallel execution scopes including start events.

As an example a start event in a multi-embedded subprocess will create a local variable in the sense that it is separate from the global scope but is still shared across all parallel execution scopes within the subprocess. It will not be local to one parallel instance.

If a local scoped variable is required to be local to a single instance of a parallel execution scope then it must be created using i/o mapping on the specific element that introduces the parallelism. Example directly on a non interrupting event sub process or a multi-embedded subprocess.

Further note if the value for such a single instance local scope variable is required to  be a snapshot of global variables at the time that a message is published then it is required to:
Create a mapping directly on the sub process element to ensure the correct scope of the element exists.
Then also pass the variable value in the published message to override the values set at creation time of the local variable as the creation time value will depend on other timings. 

Also if a variable is required only for within the subprocess scope but is required to be independent for each parallel run then it must be created outside of the subprocess scope and explicitly mapped in order to create the per parallel instance behaviour. 
This is required even if the variable is never accessed outside of the sub process scope.

**Comment by saig0** [comment]
Yes. The description is almost correct :+1: 

In other words:

If you have any kind of parallel flow and don't want to override variables then you need to scope the variables. A variable can be scoped within a variable scope by defining input variable mappings (or using multi-instance variables). A variable scope is usually an embedded subprocess, an event subprocess, or a multi-instance body.

More about variable scopes and variable propagation in the docs: https://docs.camunda.io/docs/product-manuals/concepts/variables/#variable-scopes

**Comment by npepinpe** [comment]
Let's create an issue on the docs repo which we can also add to triage.

**Comment by allanbarklie** [comment]
Thanks @saig0  and @npepinpe .

I'd be interested to know if you think here would be value in also raising a design change/suggestion.

My feeling is that if my 2 last points are correct (repeated below in italics):
_Further note if the value for such a single instance local scope variable is required to be a snapshot of global variables at the time that a message is published (which is received by a sub process) then it is required to:
Create a mapping directly on the sub process element to ensure the correct scope of the element exists.
Then also pass the variable value in the published message to override the values set at creation time of the local variable as the creation time value will depend on other timings.

Also if a variable is required only for within the subprocess scope but is required to be independent for each parallel run then it must be created outside of the subprocess scope and explicitly mapped in order to create the per parallel instance behaviour.
This is required even if the variable is never accessed outside of the sub process scope._

Then both of these can lead to awkward and error prone coding patterns that would be great to improve.
In particular when creating a new instance of an event subprocess by publishing a message.

My suggestion would be to consider changing the behaviour of I/O mapping in very limited circumstances - only changing it for Start events that are immediately enclosed by a scope that supports parallelism e.g. an event sub process or a parallel-multi-instance-sub process.  No other mappings and no other start events would be changed.

If such start events had the same behaviour as direct mappings on the the subprocess elements themselves (by which i mean if such mappings created instance specific scoped variables) then:

-  For event based subprocesses variables could be easily deterministically captured into the instance scope at the flow point time of starting the instance scope with only one mapping on the start event rather than requiring mapping on both the subprocess and the start event.
- For all such items variables could be created by start event mappings and therefore be fully encapsulated within their used scope rather than being required to leak outside to a wider scope - currently forced by the need to map on the element itself.

All other mapping behaviour could be conserved - so mappings on other elements in the subprocess would continue to behave in the current manner.  Thus I believe that all current use cases would continue to be supported e.g. being able to map to shared values across instances and also being able to map separately for instances .

I suggest this as a potential usability/error reduction improvement.
I understand there may be other reasons (beyond my knowledge) that might make this design not desirable or difficult to implement - I would just like to make the suggestion.

**Comment by saig0** [comment]
@allanbarklie thanks for your input. Feedback is always valuable :+1: 

I'm struggling to understand the problem and your suggestion completely. Please provide an example that demonstrates both. 

Just to make sure, the input or output mapping on the start event is not required. If the start event is in an event subprocess then you can limit the scope of the variable by defining the input mapping on the event subprocess. 

The event subprocess is the variable scope of the start event. An input mapping creates a local variable in the scope where it is defined. If a message start event (with or without output mappings) is completed then the message variables are [propagated](https://docs.camunda.io/docs/product-manuals/concepts/variables/#variable-propagation). This means that it updates the value of the first variable which is found in the scope hierarchy (bottom-up), or it creates a new variable in the (root) scope of the process instance.

**Comment by saig0** [comment]
I created a new issue for the docs: camunda-cloud/camunda-cloud-documentation#263

**Comment by allanbarklie** [comment]
Hi.  @saig0  thanks for raising the docs ticket.  I will try to clarify my suggestion.
Sorry for the delay, I had a deadline to meet (which your help so far has made a success).
 
A simplified use case:

A workflow manages a list of data items.
It is required that this list can be appended to.
Various things might append to that list at various times.
There is a main workflow designed to achieve this and it has a single service task where each append takes place.

Each time the data is appended to it is desired to output a report derived from the state of the list with the new append included and no other appends applied.

To achieve this the  workflow publishes a message immediately following the append to a non interrupting event sub process, sending it a copy (snapshot) of the list in the correct post append state.
The sub process produces the report in its own time.

The spacing between appends may be smaller or greater than the time taken to produce a report.
Therefore there may be several overlapping sub processes running at any time.
Each producing its own report.  
Each report showing the list in a different state.

Let's call the list, LIST
Let's call the post append copy LIST_SNAPSHOT.

To achieve this currently we need to.

1) Configure an I/O mapping on the event sub process for the variable LIST_SNAPSHOT - this is so that each instance of the sub process can have a separate LIST_SNAPSHOT variable instance.

2) Pass the correct value for LIST_SNAPSHOT  in the published message that starts the event sub process - this is so the correct value is passed.

I am currently finding (1) awkward.  There is no root variable called LIST_SNAPSHOT to use as the source for the I/O mapping.  Is there a way to declare an I/O mapping that just creates a variable from a default source value?  I've tried
 =null: LIST_SNAPSHOT  
 this passes workflow registration time checks but fails at workflow runtime.

Instead as a workaround I have created a dummy variable in the root scope, something like VALUE_NEVER_USED and then define the I/O mapping as
 =VALUE_NEVER_USED : LIST_SNAPSHOT 

Then in (2) I can pass the current value of LIST as the input value for LIST_SNAPSHOT 


Is there a misunderstanding here?  Is there something simpler I can do?


My suggestion is that instead  of the above mappings on start events for multi instance elements are changed to behave the same way as mappings on  the scope elements.
Then my implementation would be:

1) Configure an I/O mapping on the start event for sub process for the variable LIST_SNAPSHOT - this is so that each instance of the sub process can have a separate LIST_SNAPSHOT variable instance.

2) Pass the correct value for LIST_SNAPSHOT  in the published message that starts the event sub process - this is so the correct value is passed.

(1) would be easy to do as I would simply be mapping published LIST_SNAPSHOT  to local LIST_SNAPSHOT 
 =LIST_SNAPSHOT : LIST_SNAPSHOT 
There would be no awkward unused root variable.
There would be less disconnect between where I set the mapping and where I pass the value.

As a side note the current behaviour where if I do set an I/O mapping on the start event it may or may not get the value passed in the published message seems unintuitive and I'm not convinced has a valid use case.

Thanks.

**Comment by saig0** [comment]
@allanbarklie thank you for the detailed explanation :+1: 

Regarding: 

> Is there a way to declare an I/O mapping that just creates a variable from a default source value? I've tried
=null: LIST_SNAPSHOT
this passes workflow registration time checks but fails at workflow runtime.

This should work. You can define an input mapping to create a variable with a static (default) value. 

```
LIST_SNAPSHOT
= null
```

![image](https://user-images.githubusercontent.com/4305769/121654246-d197e100-ca9d-11eb-8a48-5377c89d4506.png)

Please share the failure if it doesn't work. 

---

To sum it up, you suggest defining the input mapping of `LIST_SNAPSHOT` on the message start event instead of the event subprocess. Right?

I see that it would be more intuitive to declare the mapping on the start event. This is the place where the value comes into the process. 

But in order to have the same behavior of output mappings for all elements, we would need a way to define the target scope of the variable. Currently, it propagates the variables of the output mapping to the higher scopes.

**Comment by npepinpe** [comment]
Is there anything more to add here? I will close this for now, but feel free to keep the discussion going. However, the forums may be a better place for discussions :slightly_smiling_face:
